File: backend/app/api/v1/shops.py
"""
Shop management API endpoints.
Handles shop CRUD operations.
"""
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session
from uuid import UUID
from datetime import datetime
import uuid

from app.db.session import get_db
from app.models.user import User
from app.models.shop import Shop
from app.utils.auth import get_current_active_user, require_role
from app.schemas.shop import (
    ShopCreate,
    ShopUpdate,
    ShopResponse,
    ShopListResponse
)

router = APIRouter()


@router.post("", response_model=ShopResponse, status_code=status.HTTP_201_CREATED)
async def create_shop(
    shop_data: ShopCreate,
    current_user: User = Depends(require_role(["admin"])),
    db: Session = Depends(get_db)
):
    """
    Create a new shop.
    Requires admin role.
    
    Args:
        shop_data: Shop creation data
        current_user: Current authenticated user (with required role)
        db: Database session
        
    Returns:
        Created shop
    """
    new_shop = Shop(
        id=uuid.uuid4(),
        name=shop_data.name,
        address=shop_data.address,
        timezone=shop_data.timezone,
        currency=shop_data.currency,
        gst_number=shop_data.gst_number,
        owner_user_id=None  # Can be set later
    )
    
    db.add(new_shop)
    db.commit()
    db.refresh(new_shop)
    
    return new_shop


@router.get("", response_model=ShopListResponse)
async def list_shops(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    search: Optional[str] = Query(None),
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """
    List all shops.
    Admin sees all shops, others see only their shop.
    
    Args:
        skip: Number of records to skip
        limit: Maximum number of records to return
        search: Search in shop name
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        List of shops with pagination
    """
    query = db.query(Shop)
    
    # Non-admin users can only see their own shop
    if current_user.role != 'admin':
        if not current_user.shop_id:
            return ShopListResponse(shops=[], total=0, skip=skip, limit=limit)
        query = query.filter(Shop.id == current_user.shop_id)
    
    # Search filter
    if search:
        query = query.filter(Shop.name.ilike(f"%{search}%"))
    
    total = query.count()
    shops = query.offset(skip).limit(limit).all()
    
    return ShopListResponse(
        shops=shops,
        total=total,
        skip=skip,
        limit=limit
    )


@router.get("/{shop_id}", response_model=ShopResponse)
async def get_shop(
    shop_id: UUID,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """
    Get shop details.
    
    Args:
        shop_id: Shop ID
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        Shop details
    """
    shop = db.query(Shop).filter(Shop.id == shop_id).first()
    
    if not shop:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Shop not found"
        )
    
    # Check permissions
    if current_user.role != 'admin' and current_user.shop_id != shop_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You don't have permission to access this shop"
        )
    
    return shop


@router.patch("/{shop_id}", response_model=ShopResponse)
async def update_shop(
    shop_id: UUID,
    shop_data: ShopUpdate,
    current_user: User = Depends(require_role(["owner", "admin"])),
    db: Session = Depends(get_db)
):
    """
    Update shop details.
    Requires owner or admin role.
    
    Args:
        shop_id: Shop ID
        shop_data: Shop update data
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        Updated shop
    """
    shop = db.query(Shop).filter(Shop.id == shop_id).first()
    
    if not shop:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Shop not found"
        )
    
    # Check permissions
    if current_user.role != 'admin' and current_user.shop_id != shop_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You don't have permission to update this shop"
        )
    
    # Update fields
    if shop_data.name is not None:
        shop.name = shop_data.name
    if shop_data.address is not None:
        shop.address = shop_data.address
    if shop_data.timezone is not None:
        shop.timezone = shop_data.timezone
    if shop_data.currency is not None:
        shop.currency = shop_data.currency
    if shop_data.gst_number is not None:
        shop.gst_number = shop_data.gst_number
    
    shop.updated_at = datetime.utcnow()
    db.commit()
    db.refresh(shop)
    
    return shop


@router.delete("/{shop_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_shop(
    shop_id: UUID,
    current_user: User = Depends(require_role(["admin"])),
    db: Session = Depends(get_db)
):
    """
    Delete a shop.
    Requires admin role.
    Warning: This will cascade delete all related data!
    
    Args:
        shop_id: Shop ID
        current_user: Current authenticated user
        db: Database session
    """
    shop = db.query(Shop).filter(Shop.id == shop_id).first()
    
    if not shop:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Shop not found"
        )
    
    db.delete(shop)
    db.commit()
    
    return None

File: j-backend/src/routes/v1/shops.js
/**
 * Shops API routes.
 */
const express = require('express');
const asyncHandler = require('express-async-handler');
const { v4: uuidv4 } = require('uuid');
const { authenticate, requireRole } = require('../../middleware/auth');
const { Shop } = require('../../models');
const { Op } = require('sequelize');

const router = express.Router();

// POST / - Create shop
router.post('', authenticate, requireRole(['admin']), asyncHandler(async (req, res) => {
  const { name, address, timezone = 'UTC', currency = 'INR', gst_number } = req.body;
  
  const newShop = await Shop.create({
    id: uuidv4(),
    name,
    address,
    timezone,
    currency,
    gst_number,
    owner_user_id: null,
  });
  
  res.status(201).json(newShop);
}));

// GET / - List shops
router.get('', authenticate, asyncHandler(async (req, res) => {
  const { skip = 0, limit = 100, search } = req.query;
  
  const where = {};
  
  // Non-admin users can only see their own shop
  if (req.user.role !== 'admin') {
    if (!req.user.shop_id) {
      return res.json({ shops: [], total: 0, skip: parseInt(skip, 10), limit: parseInt(limit, 10) });
    }
    where.id = req.user.shop_id;
  }
  
  if (search) {
    where.name = { [Op.iLike]: `%${search}%` };
  }
  
  const { count, rows } = await Shop.findAndCountAll({
    where,
    offset: parseInt(skip, 10),
    limit: parseInt(limit, 10),
  });
  
  res.json({
    shops: rows,
    total: count,
    skip: parseInt(skip, 10),
    limit: parseInt(limit, 10),
  });
}));

// GET /:shop_id
router.get('/:shop_id', authenticate, asyncHandler(async (req, res) => {
  const shop = await Shop.findByPk(req.params.shop_id);
  
  if (!shop) {
    return res.status(404).json({ detail: 'Shop not found' });
  }
  
  if (req.user.role !== 'admin' && req.user.shop_id !== req.params.shop_id) {
    return res.status(403).json({
      detail: "You don't have permission to access this shop",
    });
  }
  
  res.json(shop);
}));

// PATCH /:shop_id
router.patch('/:shop_id', authenticate, requireRole(['owner', 'admin']), asyncHandler(async (req, res) => {
  const shop = await Shop.findByPk(req.params.shop_id);
  
  if (!shop) {
    return res.status(404).json({ detail: 'Shop not found' });
  }
  
  if (req.user.role !== 'admin' && req.user.shop_id !== req.params.shop_id) {
    return res.status(403).json({
      detail: "You don't have permission to update this shop",
    });
  }
  
  const { name, address, timezone, currency, gst_number } = req.body;
  const updateData = {};
  
  if (name !== undefined) updateData.name = name;
  if (address !== undefined) updateData.address = address;
  if (timezone !== undefined) updateData.timezone = timezone;
  if (currency !== undefined) updateData.currency = currency;
  if (gst_number !== undefined) updateData.gst_number = gst_number;
  
  await shop.update(updateData);
  
  res.json(shop);
}));

// DELETE /:shop_id
router.delete('/:shop_id', authenticate, requireRole(['admin']), asyncHandler(async (req, res) => {
  const shop = await Shop.findByPk(req.params.shop_id);
  
  if (!shop) {
    return res.status(404).json({ detail: 'Shop not found' });
  }
  
  await shop.destroy();
  
  res.status(204).send();
}));

module.exports = router;

